File: YOLOv8_Integration/camera_publisher.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

class CameraPublisher {
public:
    CameraPublisher() {
        ros::NodeHandle nh;
        image_transport::ImageTransport it(nh);
        image_pub = it.advertise("camera/image", 10);
        timer = nh.createTimer(ros::Duration(1.0 / 5), &CameraPublisher::timerCallback, this); // Timer callback every 5 seconds
        camera_id = "/dev/video2";
        video_capture.open(camera_id, cv::CAP_V4L2);
        if (!video_capture.isOpened()) {
            ROS_ERROR("Unable to open camera");
            return;
        }
    }

    ~CameraPublisher() {
        video_capture.release();
    }

private:
    ros::Timer timer;
    image_transport::Publisher image_pub;
    cv::VideoCapture video_capture;
    std::string camera_id;

    void timerCallback(const ros::TimerEvent& event) {
        cv::Mat frame;
        if (video_capture.isOpened()) {
            if (video_capture.read(frame)) {
                sensor_msgs::ImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", frame).toImageMsg();
                image_pub.publish(msg);
            } else {
                ROS_WARN("Failed to read frame from the camera.");
            }
        } else {
            ROS_ERROR("Unable to open camera");
        }
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "usb_camera_publisher");
    CameraPublisher camera_publisher;
    ros::spin();
    return 0;
}

File: YOLOv8_Integration/camera_publisher.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2

class CameraPublisher:
    def __init__(self):
        rospy.init_node('usb_camera_publisher', anonymous=True)
        self.window_title = "USB Camera"
        self.camera_id = "/dev/video2"
        self.video_capture = cv2.VideoCapture(self.camera_id, cv2.CAP_V4L2)
        self.bridge = CvBridge()


        self.image_pub = rospy.Publisher('camera/image', Image, queue_size=10)

        #if not using callbacks 
        #self.rate = rospy.Rate(1) #Rate of 1Hz, frames would only be published at a rate of 1Hz 
        #self.publish_frames()

        self.timer = rospy.Timer(rospy.Duration(1.0 / 5), self.timer_callback)  # Timer callback every 5 seconds


    def timer_callback(self, event):
        if self.video_capture.isOpened():
            ret_val, frame = self.video_capture.read()
            if ret_val:
                image_msg = self.bridge.cv2_to_imgmsg(frame, "bgr8")
                self.image_pub.publish(image_msg)
            else:
                rospy.logwarn("Failed to read frame from the camera.")
        else:
            rospy.logerr("Unable to open camera")


if __name__ == '__main__':
    try:
        CameraPublisher()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass
